### PR TITLE
Refactored Menu scene

### PR DESCRIPTION
**Pull request by yds05238**
## Overview

- Refactored the menu scene code into separate model classes
- Created TutorialMenu, SettingsMenu, JoinMenu, LobbyMenu, and MainMenu classes for each menu screens
- Moved menu scene states into MenuState header file
- Added settings saving across gameplays

### CODE REVIEW

File: source/CIJoinMenu.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+#pragma mark -
+#pragma mark Menu Monitoring
+/**
+ * Sets whether the join game menu is currently active and visible.
+ *
+ * @param value     Whether the join game menu is currently active and visible
+ */
+void JoinMenu::setDisplay(bool value) {
```
Comment (wlong123) [inline source/CIJoinMenu.cpp:89]: NIT: can you rename value to something a bit more specific

File: source/CIJoinMenu.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+ * screen is put on display on states transitioning into JoinRoom.
+ * Screen is taken down once menu state exits JoinRoom.
+ */
+void JoinMenu::update(MenuState& state, string& joingame) {
+    if (_layer == nullptr) {
+        return;
+    }
+    // handle JoinRoom menu
+    switch (state) {
+    case MenuState::MainToJoin:
```
Comment (wlong123) [inline source/CIJoinMenu.cpp:118]: NIT: can you fix the indentation?

File: source/CIMenuScene.cpp [hunk 3]
```diff
@@ -52,185 +55,104 @@ bool MenuScene::init(const std::shared_ptr<AssetManager>& assets) {
...
+    _settings = SettingsMenu::alloc(_assets);
+    _settings->setDisplay(false);
+
+    _join = JoinMenu::alloc(_assets);
+    _join->setDisplay(false);
+
+    _lobby = LobbyMenu::alloc(_assets);
+    _lobby->setDisplay(false);
+
+    addChild(_mainmenu->getLayer(), 3);
```
Comment (wlong123) [inline source/CIMenuScene.cpp:111]: Is there a reason you put 3 here? I don't think giving all the children the same z-value does anything

**Review by wlong123 — COMMENTED** [review]
Overall looks great! Just a few minor things but this looks much cleaner than the menu scene that was all in one file

File: assets/json/menu.json [hunk 4]
```diff
@@ -532,16 +544,36 @@
...
-          "comment": "The label at the bottom-right for opponent player's info",
+        "volumeinput": {
+          "comment": "The volume slider in Settings.",
+          "type": "Slider",
+          "data": {
+            "bounds": [40, 0, 300, 0],
+            "range": [1, 100],
+            "value": 50,
+            "tick": 10,
+            "snap": true,
```
Comment (steiner26) [inline assets/json/menu.json:559]: See my other comment for explanation, but if we could make this have range [0, 1] with default value of 0.5 that would be great. Also, I feel like the snapping isn't really needed for this slider (and I guess the tick as well).

**Review by steiner26 — APPROVED** [review]
Looks good to me, I just need a few adjustments to the volume slider to make it work with the way cugl handles audio volume

**Pull request merged**